test/extensions/stats_sinks/hystrix/hystrix_test.cc

```diff
@@ -0,0 +1,338 @@
+    cluster1_.cluster_info_->stats().upstream_rq_timeout_.add(3);
+    cluster1_.cluster_info_->stats().upstream_rq_pending_overflow_.add(8);
+    sink_->flush(source_);
+  }
+
+  std::string rolling_map = sink_->printRollingWindows();
+  EXPECT_NE(std::string::npos, rolling_map.find(cluster1_name_ + ".total"));
+
+  std::string data_message = TestUtility::bufferToString(buffer);
+
+  // Check stream format and data.
+  EXPECT_EQ(getStreamField(data_message, "errorCount"), "140"); // Note that on regular operation,
+                                                                // 5xx and timeout are raised
+                                                                // together, so timeouts are reduced
+                                                                // from 5xx count
+  EXPECT_EQ(getStreamField(data_message, "requestCount"), "320");
+  EXPECT_EQ(getStreamField(data_message, "rollingCountSemaphoreRejected"), "80");
+  EXPECT_EQ(getStreamField(data_message, "rollingCountSuccess"), "70");
+  EXPECT_EQ(getStreamField(data_message, "rollingCountTimeout"), "30");
+  EXPECT_EQ(getStreamField(data_message, "errorPercentage"), "78");
+
+  // Check the values are reset.
+  buffer.drain(buffer.length());
+  sink_->resetRollingWindow();
+  sink_->flush(source_);
+  data_message = TestUtility::bufferToString(buffer);
+  validateAllZero(data_message);
+}
+
+//
+TEST_F(HystrixSinkTest, Disconnect) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  std::string data_message = TestUtility::bufferToString(buffer);
+  EXPECT_EQ(getStreamField(data_message, "rollingCountSuccess"), "0");
+  EXPECT_NE(buffer.length(), 0);
+
+  // Disconnect.
+  buffer.drain(buffer.length());
+  sink_->unregisterConnection(&callbacks_);
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Reconnect.
+  buffer.drain(buffer.length());
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  data_message = TestUtility::bufferToString(buffer);
+  EXPECT_EQ(getStreamField(data_message, "rollingCountSuccess"), "0");
+  EXPECT_NE(buffer.length(), 0);
+}
+
+TEST_F(HystrixSinkTest, AddAndRemoveClusters) {
+  InSequence s;
+
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+
+  // New cluster.
+  const std::string cluster2_name{"test_cluster2"};
+  ClusterTestInfo cluster2(cluster2_name);
+
+  // Add cluster.
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+  cluster_map_.emplace(cluster2_name, cluster2.cluster_);
+  // Redefining since cluster_map_ is returned by value.
+  ON_CALL(cluster_manager_, clusters()).WillByDefault(Return(cluster_map_));
+
+  // Generate data to both clusters.
+  sink_->flush(source_);
+  for (int i = 0; i < 12; i++) {
```

Comment:
is 12 arbitrary? I feel like some of the golden values you are comparing to below are somehow related to the constants here, but it's not super-obvious what the relationship is.  E.g. I see a few cases throughout this file where rollingCountSemaphoreRejected is tested against some expected constant, but I can't tell why that constant is expected.  E.g. why is errorCount 160?

Can we at least have comments explaining the relationship?

Reply:
Some of the values are arbitrary. We set some values to simulate the statistics and then we calculate the expected values in the stream. I was thinking we could do that in a function that will set the counter values and then calculate expected results and check them, this will make things more obvious. This function can be used in many places throughout the test. But again, I have the same concern that when the function fails, it will be hard to debug.

Reply:
I'm not following you why failures in a function are hard to debug.

Do you mean that if there's an EXPECT failure in a helper function, you might not see the whole call-stack in a test log?  I agree it's awesome when you can diagnose and fix failures purely based on a test log. However, I don't think it justifies flattening and inlining what logically should be functions just to get there.

You can also make test logs directly actionable by passing enough context to the function to print, but usually the clutter is not worth it.

And you *can* actually debug tests with gdb, in which case a function does not present any barrier to debuggability.  Note that debugging envoy unit tests with gdb is not super-obvious even if you are otherwise familiar with gdb, but I do it all the time:
```
tools/bazel-test-gdb //test/common/common:utility_test -c dbg
```

And the high order bit here, is: this test is filled with magic constants and it's not clear why they are the right ones.

Reply:
ok, I'll put them in functions and then we'll look and see if it can be improved for debugging.